### PR TITLE
add optional 'getProjectVersion' method to perform fast up-to-date ch…

### PR DESCRIPTION
**Pull request by vladima**
…ecks, if this functionality is supported by the host

### CODE REVIEW

File: src/services/services.ts [hunk 1]
```diff
@@ -2392,6 +2394,18 @@ module ts {
         }
 
         function synchronizeHostData(): void {
+            // perform fast check if host supports it
+            if (host.getProjectVersion) {
+                let hostProjectVersion = host.getProjectVersion();
+                if (hostProjectVersion) {
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: `hostProjectVersion` can be `""`, does that matter?

Comment (DanielRosenwasser) [inline src/services/services.ts]: Also, can you just do

```TypeScript
let hostProjectVersion = host.getProjectVersion && host.getProjectVersion();
```

?

Comment (vladima) [inline src/services/services.ts]: I can but don't think that squeezing a few lines of code into one line buys us much. Also I'd like to be explicit about what is happening

Comment (vladima) [inline src/services/services.ts]: >hostProjectVersion can be "", does that matter?

no, I think it is safe to assume that valid version should be non-empty string

Comment (DanielRosenwasser) [inline src/services/services.ts]: > no, I think it is safe to assume that valid version should be non-empty string

If this is the case, please document it.

**Comment by CyrusNajmabadi** [comment]
:+1:

**Comment by mhegazy** [comment]
What does projectVersion mean? when does it get updated? is changing a file leads to updating the projectVersion?

**Comment by vladima** [comment]
The assumption is that this value is changed when anything in the project changes. The capability itself is optional so if host does not expose this method or returns `undefined` - we'll fall back to the regular up-to-date check using version comparison. However if it is supported (i.e. VS can do this) - it can be used for fast up-to-date check since there is no need to fetch files\version to the script side. On my machine successful up-to-date check using version took ~35ms per file when opening large solution (~550 files) - total ~19s  of hammering CPU with no particular need. When using project version the cost of since up-to-date check dropped almost to zero.

**Pull request merged**